File: test-vectors/validate-ve-occurrence-id-v0.1.py
#!/usr/bin/env python3

WIDTH = 32


def canonical_occurrence_id(payload: bytes) -> bytes:
    if not isinstance(payload, bytes) or len(payload) != WIDTH:
        raise ValueError("OccurrenceId payload must be exactly 32 octets")
    return bytes([0x58, WIDTH]) + payload


class Rejection(Exception):
    def __init__(self, code: str):
        self.code = code
        super().__init__(code)


def reject(code: str):
    raise Rejection(code)


def decode_occurrence_id(encoded: bytes) -> bytes:
    if not isinstance(encoded, bytes) or not encoded:
        reject("missing-value")
    initial = encoded[0]
    if initial & 0xE0 == 0xC0:
        reject("tag-forbidden")
    if initial == 0x5F:
        reject("indefinite-length")
    if initial & 0xE0 != 0x40:
        reject("wrong-cbor-type")
    if initial != 0x58:
        if initial in (0x59, 0x5A, 0x5B):
            reject("noncanonical-length")
        reject("unsupported-byte-string-form")
    if len(encoded) < 2:
        reject("truncated-length")
    if encoded[1] != WIDTH:
        reject("invalid-length")
    encoded_length = 2 + WIDTH
    if len(encoded) < encoded_length:
        reject("truncated-payload")
    if len(encoded) > encoded_length:
        reject("trailing-bytes")
    payload = encoded[2:]
    if canonical_occurrence_id(payload) != encoded:
        reject("noncanonical-encoding")
    return payload


def accepted(name: str, encoded: bytes):
    payload = decode_occurrence_id(encoded)
    if len(payload) != WIDTH:
        raise AssertionError(f"{name}: wrong decoded width")


def rejected(name: str, encoded: bytes, expected: str):
    try:
        decode_occurrence_id(encoded)
    except Rejection as error:
        if error.code == expected:
            return
        raise AssertionError(f"{name}: expected {expected}, got {error.code}") from error
    raise AssertionError(f"{name}: expected rejection")


zero = bytes(WIDTH)
nonzero = bytes([0xA5]) * WIDTH
one_octet_different = bytes([0xA5]) * (WIDTH - 1) + b"\xA4"

# A1–A4
accepted("A1", canonical_occurrence_id(zero))
accepted("A2", canonical_occurrence_id(nonzero))
accepted("A3", canonical_occurrence_id(nonzero))
assert decode_occurrence_id(canonical_occurrence_id(nonzero)) == decode_occurrence_id(canonical_occurrence_id(nonzero)), "A3: equal payloads must compare equal"
assert decode_occurrence_id(canonical_occurrence_id(nonzero)) != decode_occurrence_id(canonical_occurrence_id(one_octet_different)), "A4: distinct payloads must compare unequal"

# R1–R8
rejected("R1", bytes([0x58, 0x1F]) + bytes(31), "invalid-length")
rejected("R2", bytes([0x58, 0x21]) + bytes(33), "invalid-length")
rejected("R3", bytes([0x78, 0x20]) + bytes([0x61]) * 32, "wrong-cbor-type")
rejected("R4", bytes([0x00]), "wrong-cbor-type")
rejected("R5", bytes([0x5F, 0x58, 0x20]) + zero + bytes([0xFF]), "indefinite-length")
rejected("R6", bytes([0xC0]) + canonical_occurrence_id(zero), "tag-forbidden")
rejected("R7", bytes([0x59, 0x00, 0x20]) + zero, "noncanonical-length")
rejected("R8", canonical_occurrence_id(zero) + bytes([0x00]), "trailing-bytes")

print("VE OccurrenceId v0.1 vectors passed: 4 accepted, 8 rejected.")

File: test-vectors/validate-ve-occurrence-id-v0.1.mjs
#!/usr/bin/env node

const WIDTH = 32;

function bytesFromHex(value) {
  if (!/^[0-9a-f]*$/i.test(value) || value.length % 2 !== 0) throw new Error("fixture hex is malformed");
  return Buffer.from(value, "hex");
}

function canonicalOccurrenceId(payload) {
  if (!Buffer.isBuffer(payload) || payload.length !== WIDTH) throw new Error("OccurrenceId payload must be exactly 32 octets");
  return Buffer.concat([Buffer.from([0x58, WIDTH]), payload]);
}

function invalid(code) {
  const error = new Error(code);
  error.code = code;
  throw error;
}

function decodeOccurrenceId(input) {
  if (!Buffer.isBuffer(input) || input.length === 0) invalid("missing-value");
  const initial = input[0];
  if ((initial & 0xe0) === 0xc0) invalid("tag-forbidden");
  if (initial === 0x5f) invalid("indefinite-length");
  if ((initial & 0xe0) !== 0x40) invalid("wrong-cbor-type");
  if (initial !== 0x58) {
    if (initial === 0x59 || initial === 0x5a || initial === 0x5b) invalid("noncanonical-length");
    invalid("unsupported-byte-string-form");
  }
  if (input.length < 2) invalid("truncated-length");
  if (input[1] !== WIDTH) invalid("invalid-length");
  const encodedLength = 2 + WIDTH;
  if (input.length < encodedLength) invalid("truncated-payload");
  if (input.length > encodedLength) invalid("trailing-bytes");
  const payload = input.subarray(2);
  if (!canonicalOccurrenceId(payload).equals(input)) invalid("noncanonical-encoding");
  return payload;
}

function assertAccepted(name, input) {
  const payload = decodeOccurrenceId(input);
  if (payload.length !== WIDTH) throw new Error(`${name}: wrong decoded width`);
}

function assertRejected(name, input, code) {
  try {
    decodeOccurrenceId(input);
  } catch (error) {
    if (error.code === code) return;
    throw new Error(`${name}: expected ${code}, got ${error.code ?? error.message}`);
  }
  throw new Error(`${name}: expected rejection`);
}

const zero = Buffer.alloc(WIDTH, 0x00);
const nonzero = Buffer.alloc(WIDTH, 0xa5);
const oneOctetDifferent = Buffer.concat([Buffer.alloc(WIDTH - 1, 0xa5), Buffer.from([0xa4])]);

// A1–A4
assertAccepted("A1", canonicalOccurrenceId(zero));
assertAccepted("A2", canonicalOccurrenceId(nonzero));
assertAccepted("A3", canonicalOccurrenceId(nonzero));
if (!decodeOccurrenceId(canonicalOccurrenceId(nonzero)).equals(decodeOccurrenceId(canonicalOccurrenceId(nonzero)))) throw new Error("A3: equal payloads must compare equal");
if (decodeOccurrenceId(canonicalOccurrenceId(nonzero)).equals(decodeOccurrenceId(canonicalOccurrenceId(oneOctetDifferent)))) throw new Error("A4: distinct payloads must compare unequal");

// R1–R8
assertRejected("R1", Buffer.concat([Buffer.from([0x58, 0x1f]), Buffer.alloc(31)]), "invalid-length");
assertRejected("R2", Buffer.concat([Buffer.from([0x58, 0x21]), Buffer.alloc(33)]), "invalid-length");
assertRejected("R3", Buffer.concat([Buffer.from([0x78, 0x20]), Buffer.alloc(32, 0x61)]), "wrong-cbor-type");
assertRejected("R4", bytesFromHex("00"), "wrong-cbor-type");
assertRejected("R5", Buffer.concat([Buffer.from([0x5f, 0x58, 0x20]), zero, Buffer.from([0xff])]), "indefinite-length");
assertRejected("R6", Buffer.concat([Buffer.from([0xc0]), canonicalOccurrenceId(zero)]), "tag-forbidden");
assertRejected("R7", Buffer.concat([Buffer.from([0x59, 0x00, 0x20]), zero]), "noncanonical-length");
assertRejected("R8", Buffer.concat([canonicalOccurrenceId(zero), Buffer.from([0x00])]), "trailing-bytes");

console.log("VE OccurrenceId v0.1 vectors passed: 4 accepted, 8 rejected.");
